llvm/lib/Target/RISCV/RISCVMacroFusion.cpp

```diff
@@ -27,29 +111,18 @@ static bool isLUIADDI(const MachineInstr *FirstMI,
   if (SecondMI.getOpcode() != RISCV::ADDI &&
       SecondMI.getOpcode() != RISCV::ADDIW)
     return false;
-
   // Assume the 1st instr to be a wildcard if it is unspecified.
   if (!FirstMI)
     return true;
 
   if (FirstMI->getOpcode() != RISCV::LUI)
     return false;
 
-  Register FirstDest = FirstMI->getOperand(0).getReg();
-
-  // Destination of LUI should be the ADDI(W) source register.
-  if (SecondMI.getOperand(1).getReg() != FirstDest)
+  // The first operand of ADDI might be a frame index.
```

Comment:
This check was removed previously because it wasn't needed. Was this a bad merge or is it needed now?

Reply:
In this code:

```
$rd0 =  lui $imm0
$rd1 = addi $rs0, $imm1
```

In order for fusion to happen it must be that `$rd0 == rd1` and `$rd0 == rs0`. Same for other fusions. Both checks are done in `checkRegisters` now. I think both checks were needed before two, according to this comment:

```
// Fuse LUI followed by ADDI or ADDIW.
// rd = imm[31:0] which decomposes to
// lui rd, imm[31:12]
// addi(w) rd, rd, imm[11:0]

```

Reply:
I was specifically refering to the FrameIndex comment.

Reply:
I think we do need the check `SecondMI.getOperand(1).isReg()`  because later we do `SecondMI.getOperand(1).getReg() != FirstDest` - this may crash, right?

I've put that check as a first line of `checkRegisters`.

Reply:
As was explained in #68701, if the ADDI operand isn't a register then there can be no data dependency between the LUI and the ADDi so it wouldn't be considered for macrofusion.

Reply:
oh, sorry I didn't realize that by the time we call `shouldScheduleAdjacent` we know that there is a data dependency. Ok, I looked at  `MacroFusion.cpp : 177` - we only consider pairs of vertices connected by an edge of a schedule dag. But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

Reply:
> But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

Yes, I have to say that it's possible to have other order dependencies like `Artificial`. For example, `BaseMemOpClusterMutation::clusterNeighboringMemOps`. Though we may not add artificial edges (not for fusion) between ALU instructions. I haven't run into any problem after remove the `isReg()` yet, but I think we may add it back since there are some potential risks of assertion theoretically.